Autocompletion for `return new …` statement differs between `Task` and `ValueTask` methods

**Version Used**: VS 2022 17.3.6

**Steps to Reproduce**:

1. Paste the code below into the editor.
2. Invoke the autocompletion list after each of the two `new` keywords.

```csharp
class C
{
    async Task<string> M1Async()
    {
        await Task.CompletedTask;
        return new ;
    }

    async ValueTask<string> M2Async()
    {
        await Task.CompletedTask;
        return new ;
    }
}
```

**Expected Behavior**:
In both cases `string` is suggested, as the methods are marked as `async`.

**Actual Behavior**:
* `object` is suggested in the `Task` method.  
  ![grafik](https://user-images.githubusercontent.com/823398/196961513-b5b63599-85b1-4658-a6d3-d7ebc8d1a5f5.png)
* `ValueTask<object>` is suggested in the `ValueTask` method.  
  ![grafik](https://user-images.githubusercontent.com/823398/196961653-ca0fabdb-94a8-465e-ac52-7bd61dd203aa.png)

Accepting the autocompletion of `ValueTask<string>` leads to error CS4016.

--- a/src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectCreationCompletionProviderTests.cs
+++ b/src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectCreationCompletionProviderTests.cs
@@ -100,6 +100,23 @@ class Program
         }
 
         [Fact]
+        public async Task InAsyncMethodReturnValueTask()
+        {
+            var markup =
+@"using System;
+using System.Threading.Tasks;
+
+class Program
+{
+    async ValueTask&lt;string&gt; M2Async()
+    {
+        return new $$;
+    }
+}";
+            await VerifyItemExistsAsync(MakeMarkup(markup), "string");
+        }
+
+        [Fact]
         public async Task IsCommitCharacterTest()
         {
             const string markup = @"
@@ -768,6 +785,19 @@ class Program
     private List<int> o = new List<int>();
 }";
             await VerifyProviderCommitAsync(markup, "List<int>", expectedMark, commitChar: ';');
+        }
+
+        private static string MakeMarkup(string source, string languageVersion = "Preview")
+        {
+            return $$"""
+<Workspace>
+    <Project Language="C#" AssemblyName="Assembly" CommonReferencesNet6="true" LanguageVersion="{{languageVersion}}">
+        <Document FilePath="Test.cs">
+{{source}}
+        </Document>
+    </Project>
+</Workspace>
+""";
         }
     }
 }

--- a/src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb
+++ b/src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb
@@ -11066,5 +11066,25 @@ public class Bar&lt;T&gt; : ISomeInterface&lt;T&gt;
                 Await state.AssertCompletionItemsContain("SomeExtMethod", displayTextSuffix:="<>")
             End Using
         End Function
+
+        <WpfFact, WorkItem(64862, "https://github.com/dotnet/roslyn/issues/64862")>
+        Public Async Function TestAsyncMethodReturningValueTask() As Task
+            Using state = TestStateFactory.CreateCSharpTestState(
+                <Document>
+using System.Threading.Tasks;
+
+class Program
+{
+    async ValueTask&lt;string&gt; M2Async()
+    {
+        return new $$;
+    }
+}
+                </Document>)
+
+                state.SendInvokeCompletionList()
+                Await state.AssertCompletionItemsContain("string", displayTextSuffix:="")
+            End Using
+        End Function
     End Class
 End Namespace

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs
@@ -2026,13 +2026,13 @@ namespace Microsoft.CodeAnalysis.CSharp
             {
                 if (isAsync)
                 {
-                    if (type.OriginalDefinition.Equals(this.Compilation.TaskOfTType()))
+                    if (type.OriginalDefinition.Equals(this.Compilation.TaskOfTType()) || type.OriginalDefinition.Equals(this.Compilation.ValueTaskOfTType()))
                     {
                         var namedTypeSymbol = (INamedTypeSymbol)type;
                         return namedTypeSymbol.TypeArguments[0];
                     }
 
-                    if (type.OriginalDefinition.Equals(this.Compilation.TaskType()))
+                    if (type.OriginalDefinition.Equals(this.Compilation.TaskType()) || type.OriginalDefinition.Equals(this.Compilation.ValueTaskType()))
                     {
                         return this.Compilation.GetSpecialType(SpecialType.System_Void);
                     }

